### PR TITLE
[Quant][core][gpu][improvement] Refactored implementation for conv2d_cudnn to use packed parameters

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73510

Summary:
The previous implementation introduced in https://github.com/pytorch/pytorch/pull/70622
and expanded on in https://github.com/pytorch/pytorch/pull/72770,
https://github.com/pytorch/pytorch/pull/73035, https://github.com/pytorch/pytorch/pull/73337
did not make use of packed parameters. This PR refactors the existing
implementation to use packed parameters for cudnn conv2d in the same manner
as was done for qnnpack and fbgemm in the following files:
aten/src/ATen/native/quantized/cpu/fbgemm_utils.h.
aten/src/ATen/native/quantized/cpu/qnnpack_utils.h.
aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp.
aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp (note this file will be
refactored into two files (one located in /quantized/ and the other in
/quantized/cpu/) in a subsequent PR, as we are currently using the dispatch introduced
in this file for the cudnn operator as well)

This allows for all cudnn operators to be registered as quantized::conv2d,
quantized::conv2d_relu, quantized::conv2d_prepack, and to allow the dispatcher
to determine which backend to use (e.g., cuda/cudnn, fbgemm, or qnnpack).

Test cases were also modified to adhere to the methodology of using
prepacking the weight & bias prior to passing it into the conv2d operator.

We also ensured that the refactorization did not result in a reduction in speed
by verifying that the computation times in the benchmark test case (see test plan below)
are consistent with the results pre-refactorization.

Note the following:
apply_impl is now what was formerly raw_cudnn_convolution_forward
apply_impl_helper is now what was formerly raw_cudnn_convolution_forward_out

Test plan:
In pytorch main directory, execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```
for accuracy testing and
```
python test/test_quantization.py TestQuantizedConv.test_benchmark
```
for benchmark testing.

Differential Revision: [D34803275](https://our.internmc.facebook.com/intern/diff/D34803275)

### CODE REVIEW

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

File: aten/src/ATen/native/quantized/cudnn/Conv.cpp [hunk 1]
```diff
@@ -436,94 +427,96 @@ out_int8 = (act_fp32 * w_fp32 + [bias_fp32]) / out_scale + out_zero_point
...
   const int N = act.size(0);
   const int D = kSpatialDim == 3 ? act.size(2) : 1;
   const int H = act.size(kSpatialDim);
   const int W = act.size(kSpatialDim + 1);
-  const int M = weight.size(0); // output channels
-  std::vector<int64_t> kernel_size = {weight.size(2), weight.size(3)};
-  at::SmallVector<int64_t, kSpatialDim + 2> output_shape{MakeConvOutputShape<kSpatialDim>(N, M, {H, W},
-  kernel_size, stride, padding, dilation)};
-  Tensor quantized_output = at::_empty_affine_quantized(
+  const int M = orig_weight.size(0); // output channels
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/Conv.cpp:440]: nit: is this the saved weight? should we give it a special name to make it more explicit? like `weight_`?

File: aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      bool transpose) {
+    return PackedConvWeightCudnn<kSpatialDim>::prepack(
+        weight, bias, stride, padding, output_padding, dilation, groups,
+        transpose);
+  }
+};
+
+TORCH_LIBRARY_IMPL(quantized, QuantizedCUDA, m) {
+  // Conv
+  // conv_prepack is deprecated, please use conv2d_prepack for 2D conv.
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:143]: nit: do we need this comment?

File: aten/src/ATen/native/quantized/cudnn/Conv.cpp (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cudnn/Conv.cpp:440]: same for `bias` I think

File: aten/src/ATen/native/quantized/cudnn/cudnnpack_utils.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+  int64_t groups() const override {
+    return groups_;
+  }
+
+  bool transpose() const override {
+    return transpose_;
+  }
+
+ private:
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cudnn/cudnnpack_utils.h:98]: naming convention for private member variable is `variable_name_` I think

**Review by jerryzh168 — APPROVED** [review]
looks great!

**Comment by jerryzh168** [comment]
please add appropriate labels to the PR as well

File: aten/src/ATen/native/quantized/cudnn/Conv.cpp (hunk 1 above)
Comment (dzdang) [reply aten/src/ATen/native/quantized/cudnn/Conv.cpp:440]: Yeah orig_weight is the saved weight. qnnpack also used the same convention with no underscore. 
I can append the underscore and then have another PR that does the same for qnnpack for uniformity. 

I can do the same for `bias`. I foudn that the naming convention isn't consistent, e.g., in https://github.com/pytorch/pytorch/blob/gh/dzdang/42/head/aten/src/ATen/native/quantized/cpu/fbgemm_utils.h
you can see both `bias` (for `PackedConvWeight` class) and `bias_` (for `PackedLinearWeight` class) used for member variables

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cudnn/Conv.cpp:440]: yeah we should fix them I think

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34803275).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34803275).

**Comment by suo** [comment]
This PR broke backwards compatibility checks: https://hud.pytorch.org/pytorch/pytorch/commit/85c790ed9fc82abf70f49f6c847a72913a36d16f. @dzdang I am reverting.

**Pull request closed without merging**